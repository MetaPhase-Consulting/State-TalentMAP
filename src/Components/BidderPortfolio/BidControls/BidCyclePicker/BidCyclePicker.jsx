import React, { Component } from 'react';
import Picky from 'react-picky';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import bowser from 'bowser';
import { filter, indexOf, isArray, map, throttle } from 'lodash';
import { EMPTY_FUNCTION } from 'Constants/PropTypes';
import { bidderPortfolioSeasonsFetchData, bidderPortfolioSetSeasons } from 'actions/bidderPortfolio';
import ListItem from './ListItem';

// TODO - Running into an issue where the label/span element is also
// passing up an event almost concurrently (400ms difference).
// This only happens in IE11, so we add a leading throttle so that
// any additional events within 1000ms do not get called.
const browser = bowser.getParser(window.navigator.userAgent);
const isIE = browser.satisfies({ 'internet explorer': '<=11' });
const THROTTLE_MS = isIE ? 1000 : 0;

export function renderList({ items, ...rest }) {
  return items.map(item => <ListItem {...rest} key={item.id} item={item.description} />);
}

class BidCyclePicker extends Component {
  constructor(props) {
    super(props);
    this.state = {
      arrayValue: [],
    };
    this.selectMultipleOption = throttle(
      this.selectMultipleOption.bind(this),
      THROTTLE_MS,
      { trailing: false, leading: true },
    );
    this.bidSeasonsToIds = this.bidSeasonsToIds.bind(this);
  }
  componentWillMount() {
    // Only perform once in the session since this will rarely change.
    if (!this.props.seasons.length) {
      this.props.fetchSeasons();
    }
  }
  componentDidMount() {
    this.props.setSeasonsCb(this.getSeasons());
  }
  componentDidUpdate() {
    this.props.setSeasonsCb(this.getSeasons());
  }
  setSeasons() {
    const seasons = this.bidSeasonsToIds();
    this.props.setSeasons(seasons);
  }
  getSeasons() {
    return this.bidSeasonsToIds();
  }
  bidSeasonsToIds() {
    const { arrayValue } = this.state;
    const { seasons } = this.props;
    let ids$ = isArray(seasons) ? [...seasons] : [];
    ids$ = filter(ids$, f => indexOf(arrayValue, f.description) > -1);
    ids$ = map(ids$, m => m.id);
    return ids$;
  }
  selectMultipleOption(value) {
    this.setState({ arrayValue: value }, () => this.setSeasons());
  }
  render() {
    const { arrayValue } = this.state;
    const { seasons, isLoading, hasErrored } = this.props; // eslint-disable-line
    return (
      <div className="bid-cycle-picker-container">
        <div className="label">Bid season:</div>
        <Picky
          placeholder="Select season(s)"
          value={arrayValue}
          options={seasons}
          onChange={this.selectMultipleOption}
          numberDisplayed={1}
          multiple
          dropdownHeight={600}
          renderList={renderList}
          disabled={isLoading}
        />
      </div>
    );
  }
}

BidCyclePicker.propTypes = {
  seasons: PropTypes.arrayOf(PropTypes.shape({})),
  isLoading: PropTypes.bool,
  hasErrored: PropTypes.bool,
  fetchSeasons: PropTypes.func.isRequired,
  setSeasons: PropTypes.func.isRequired,
  setSeasonsCb: PropTypes.func,
};

BidCyclePicker.defaultProps = {
  seasons: [],
  isLoading: false,
  hasErrored: false,
  setSeasonsCb: EMPTY_FUNCTION,
};

const mapStateToProps = state => ({
  seasons: state.bidderPortfolioSeasons,
  isLoading: state.bidderPortfolioSeasonsIsLoading,
  hasErrored: state.bidderPortfolioSeasonsHasErrored,
});

export const mapDispatchToProps = dispatch => ({
  fetchSeasons: () => dispatch(bidderPortfolioSeasonsFetchData()),
  setSeasons: (arr = []) => dispatch(bidderPortfolioSetSeasons(arr)),
});

export default connect(mapStateToProps, mapDispatchToProps)(BidCyclePicker);
