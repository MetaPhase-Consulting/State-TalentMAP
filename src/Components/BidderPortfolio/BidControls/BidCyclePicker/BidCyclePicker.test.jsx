import { shallow } from 'enzyme';
import React from 'react';
import toJSON from 'enzyme-to-json';
import BidCyclePicker, { renderList } from './BidCyclePicker';

describe('BidControlsComponent', () => {
  it('is defined', () => {
    const wrapper = shallow(
      <BidCyclePicker />,
    );
    expect(wrapper).toBeDefined();
  });

  it('sets state when selectOption() is called', () => {
    const wrapper = shallow(
      <BidCyclePicker />,
    );
    const val = 1;
    wrapper.instance().selectOption(val);
    expect(wrapper.instance().state.value).toBe(val);
  });

  it('sets state when selectMultipleOption() is called', () => {
    const wrapper = shallow(
      <BidCyclePicker />,
    );
    const val = [1];
    wrapper.instance().selectMultipleOption(val);
    expect(wrapper.instance().state.arrayValue).toEqual(val);
  });

  it('returns an array of items when renderList() is called', () => {
    const output = renderList(
      { items: ['a', 'b', 'c'], selectValue: () => {}, getIsSelected: () => false },
    );
    expect(output.length).toBe(3);
  });

  it('matches snapshot', () => {
    const wrapper = shallow(
      <BidCyclePicker />,
    );
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});
