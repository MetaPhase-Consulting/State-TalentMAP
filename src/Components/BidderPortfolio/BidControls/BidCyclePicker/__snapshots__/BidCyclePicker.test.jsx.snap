// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BidControlsComponent matches snapshot 1`] = `
<div
  className="bid-cycle-picker-container usa-form"
>
  <div
    className="label"
  >
    Bid Season:
  </div>
  <r
    disabled={false}
    dropdownHeight={600}
    filterDebounce={150}
    id="picky"
    keepOpen={true}
    multiple={true}
    numberDisplayed={1}
    onChange={[Function]}
    options={Array []}
    placeholder="Select season(s)"
    renderList={[Function]}
    selectAllMode="default"
    selectAllText="Select all"
    tabIndex={0}
    value={Array []}
  />
</div>
`;
